In [2]:
import pandas as pd

# A. Introduction to Pandas DataFrame and Transformation (45pts)

## Beginner

### 1. Membuat DataFrame Manual (5pts)

Buatlah dataframe secara manual seperti gambar dibawah ini.

**Expected Result:**

![Expected Result](assets/homework_exercise_1_result.png)

In [136]:
# Nomor 1
data = {
    "name": ["Fiqry", "Iqbal", "Monica", "Rama", "Johan"],
    "age": [23, 21, 22, 24, 26],
    "phone_number": ["+62813123414", "+6287842464", "+62813125554", "+6287834464", "+62813113414"],
    "ielts_score": [6.5, None, 7.5, 6.5, 8.0]
}
df = pd.DataFrame(data)
df

name  age  phone_number  ielts_score
0   Fiqry   23  +62813123414          6.5
1   Iqbal   21   +6287842464          NaN
2  Monica   22  +62813125554          7.5
3    Rama   24   +6287834464          6.5
4   Johan   26  +62813113414          8.0

### 2. Analisa Sederhana DataFrame (10pts)

Lakukan import `application_processed.csv` data kedalam notebook (5pts) dan carilah beberapa informasi pada dataset `application_processed` (5pts) dan jawablah beberapa pertanyaan sebagai berikut:

a. Column apa saja yang memiliki nilai NaN / None?
b. Berapa rata-rata pada column `amount_income_total`?
c. Berapa median pada column `amount_income_total`?

**Hint**:
- Menggunakan operasi dasar DataFrame pada topik Introduction Pandas DataFrame and Transformation


In [6]:
## Your code below here
df2 = pd.read_csv('application_processed.csv')
df2

id  gender  flag_own_car  flag_own_property  total_children  \
0       5548815  Female             1                  1               1   
1       5692601  Female             1                  1               0   
2       5625007  Female             1                  0               0   
3       6529653    Male             0                  1               0   
4       6209935    Male             0                  1               1   
...         ...     ...           ...                ...             ...   
131562  6108756  Female             1                  1               0   
131563  5938512    Male             1                  0               0   
131564  6824835    Male             1                  0               1   
131565  5009009    Male             0                  1               0   
131566  6054185  Female             0                  1               2   

        amount_income_total           income_type  \
0                  225000.0         State servant   
1                  180000.0               Working   
2                  157500.0             Pensioner   
3                  202500.0               Working   
4                  157500.0  Commercial associate   
...                     ...                   ...   
131562             157500.0               Working   
131563             202500.0               Working   
131564             225000.0         State servant   
131565             166500.0               Working   
131566             135000.0               Working   

                       education_type family_status       housing_type  \
0       Secondary / secondary special       Married  House / apartment   
1       Secondary / secondary special       Married  House / apartment   
2       Secondary / secondary special         Widow  House / apartment   
3       Secondary / secondary special       Married  House / apartment   
4       Secondary / secondary special       Married  House / apartment   
...                               ...           ...                ...   
131562               Higher education       Married  House / apartment   
131563  Secondary / secondary special       Married  House / apartment   
131564              Incomplete higher       Married  House / apartment   
131565  Secondary / secondary special       Married  House / apartment   
131566  Secondary / secondary special       Married  House / apartment   

        days_birth  days_employed  flag_mobile_phone  flag_work_phone  \
0           -12944          -6127                  1                0   
1           -12484          -1874                  1                0   
2           -21656         365243                  1                0   
3           -10050          -2117                  1                0   
4           -10371          -1755                  1                0   
...            ...            ...                ...              ...   
131562      -10867          -2637                  1                0   
131563      -17815          -5109                  1                0   
131564      -10472          -3601                  1                0   
131565      -18926          -6276                  1                1   
131566      -11390          -2736                  1                0   

        flag_phone  flag_email occupation_type  total_family_members  
0                0           0        Laborers                   3.0  
1                0           0             NaN                   2.0  
2                0           0             NaN                   1.0  
3                0           0         Drivers                   2.0  
4                0           0         Drivers                   3.0  
...            ...         ...             ...                   ...  
131562           0           0        Managers                   2.0  
131563           0           0             NaN                   2.0  
131564           0           0  Security staff            

In [20]:
# Nomor 1
df2.isnull().sum()

id                          0
gender                      0
flag_own_car                0
flag_own_property           0
total_children              0
amount_income_total         0
income_type                 0
education_type              0
family_status               0
housing_type                0
days_birth                  0
days_employed               0
flag_mobile_phone           0
flag_work_phone             0
flag_phone                  0
flag_email                  0
occupation_type         40095
total_family_members        0
dtype: int64

### Nomor 2a
Kolom yang memiliki Nan / None adalah kolom 'occupation_type' yang berjumlah 40095 baris Nan / None 

In [37]:
# Nomor 2b
rata_rata_income = df2['amount_income_total'].mean()
print(f'Rata Rata amount_income_total: {rata_rata_income}')

Rata Rata Income: 188040.57643679646


In [38]:
# Nomor 2c
median_income = df2['amount_income_total'].median()
print(f'Median amount_income_total: {median_income}')

Median amount_income_total: 162000.0


## Intermediate

### 3. Mengolah Dataset (30pts)

Dengan menggunakan dataset `application_processed.csv` buatlah dataframe baru dengan ketentuan:
1. Ambil semua users yang memiliki `email`. (ditandai dengan nilai pada column `flag_email` = 1) dan tidak memilik data `duplicates`. (5pts)
2. Buatlah column baru bernama `has_car_and_property` dimana nilainya merupakan range dari `0-1`. `1` apabila `flag_own_car` bernilai `1` dan `flag_own_property` bernilai `1`. `0` apabila terdapat atau semua nilai pada column `flag_own_car` dan `flag_own_property` terdapat angka 0. (5pts)
3. Ambil column yang diperlukan yaitu `id`, `gender`, `has_car_and_property`, `income_type`, `education_type`, `family_status`, `housing_type`, `occupation_type`, `amount_income_total`. (5pts)
4. Lakukan filter dengan ketentuan sebagai berikut: (5pts)
    - Memiliki `car` dan `property`
    - Tidak ada nilai null pada column `occupation_type`
    - Memiliki value pada column `occupation_type` dengan pola `staff`.
5. Sudah dilakukan sorting berdasarkan `income_type` dari `A-Z` dan `amount_income_total` dari `besar ke kecil`. (5pts)
6. Simpan kedalam bentuk `.csv` dengan nama `homework_3_result.csv`. (5pts)

**Expected Result:**

![Expected Result](assets/homework_exercise_3_result.png)

In [29]:
df3 = pd.read_csv('application_processed.csv')
df3

id  gender  flag_own_car  flag_own_property  total_children  \
0       5548815  Female             1                  1               1   
1       5692601  Female             1                  1               0   
2       5625007  Female             1                  0               0   
3       6529653    Male             0                  1               0   
4       6209935    Male             0                  1               1   
...         ...     ...           ...                ...             ...   
131562  6108756  Female             1                  1               0   
131563  5938512    Male             1                  0               0   
131564  6824835    Male             1                  0               1   
131565  5009009    Male             0                  1               0   
131566  6054185  Female             0                  1               2   

        amount_income_total           income_type  \
0                  225000.0         State servant   
1                  180000.0               Working   
2                  157500.0             Pensioner   
3                  202500.0               Working   
4                  157500.0  Commercial associate   
...                     ...                   ...   
131562             157500.0               Working   
131563             202500.0               Working   
131564             225000.0         State servant   
131565             166500.0               Working   
131566             135000.0               Working   

                       education_type family_status       housing_type  \
0       Secondary / secondary special       Married  House / apartment   
1       Secondary / secondary special       Married  House / apartment   
2       Secondary / secondary special         Widow  House / apartment   
3       Secondary / secondary special       Married  House / apartment   
4       Secondary / secondary special       Married  House / apartment   
...                               ...           ...                ...   
131562               Higher education       Married  House / apartment   
131563  Secondary / secondary special       Married  House / apartment   
131564              Incomplete higher       Married  House / apartment   
131565  Secondary / secondary special       Married  House / apartment   
131566  Secondary / secondary special       Married  House / apartment   

        days_birth  days_employed  flag_mobile_phone  flag_work_phone  \
0           -12944          -6127                  1                0   
1           -12484          -1874                  1                0   
2           -21656         365243                  1                0   
3           -10050          -2117                  1                0   
4           -10371          -1755                  1                0   
...            ...            ...                ...              ...   
131562      -10867          -2637                  1                0   
131563      -17815          -5109                  1                0   
131564      -10472          -3601                  1                0   
131565      -18926          -6276                  1                1   
131566      -11390          -2736                  1                0   

        flag_phone  flag_email occupation_type  total_family_members  
0                0           0        Laborers                   3.0  
1                0           0             NaN                   2.0  
2                0           0             NaN                   1.0  
3                0           0         Drivers                   2.0  
4                0           0         Drivers                   3.0  
...            ...         ...             ...                   ...  
131562           0           0        Managers                   2.0  
131563           0           0             NaN                   2.0  
131564           0           0  Security staff            

1. Ambil semua users yang memiliki `email`. (ditandai dengan nilai pada column `flag_email` = 1) dan tidak memilik data `duplicates`. (5pts)

In [68]:
df3_with_email = df3[df3['flag_email'] == True].drop_duplicates()
df3_with_email

id  gender  flag_own_car  flag_own_property  total_children  \
5       5541414  Female             0                  1               0   
8       5169560  Female             1                  1               0   
22      5884982  Female             0                  1               0   
23      6297619    Male             1                  1               0   
28      6006818  Female             0                  1               0   
...         ...     ...           ...                ...             ...   
131511  5345596    Male             1                  1               1   
131514  5680139  Female             0                  0               1   
131518  5708288  Female             0                  1               2   
131537  5352502  Female             1                  1               0   
131555  6234604  Female             1                  1               1   

        amount_income_total           income_type  \
5                  211500.0               Working   
8                   90000.0               Working   
22                 337500.0  Commercial associate   
23                 360000.0  Commercial associate   
28                 247500.0             Pensioner   
...                     ...                   ...   
131511             112500.0               Working   
131514              67500.0               Working   
131518             135000.0               Working   
131537             103500.0  Commercial associate   
131555             171000.0               Working   

                       education_type         family_status  \
5       Secondary / secondary special        Civil marriage   
8       Secondary / secondary special               Married   
22      Secondary / secondary special               Married   
23                   Higher education  Single / not married   
28                   Higher education               Married   
...                               ...                   ...   
131511               Higher education               Married   
131514  Secondary / secondary special               Married   
131518               Higher education               Married   
131537               Higher education        Civil marriage   
131555  Secondary / secondary special               Married   

               housing_type  days_birth  days_employed  flag_mobile_phone  \
5         House / apartment      -18171          -1671                  1   
8         House / apartment      -16023          -3532                  1   
22      Municipal apartment      -15491           -703                  1   
23         Rented apartment      -15796          -1625                  1   
28        House / apartment      -23032         365243                  1   
...                     ...         ...            ...                ...   
131511    House / apartment      -17318           -246                  1   
131514    House / apartment      -15140          -2859                  1   
131518         With parents      -13296          -2178                  1   
131537    House / apartment      -11772           -113                  1   
131555     Office apartment      -13908           -993                  1   

        flag_work_phone  flag_phone  flag_email        occupation_type  \
5                     0           1           1                    NaN   
8                     0           0           1            Sales staff   
22                    0           1           1               Managers   
23                    0           0           1  High skill tech staff   
28                    0           0           1                    NaN   
...                 ...         ...         ...                    ...   
131511                0           0           1               Managers   
131514                0           0           1          Cooking staff   
131518                0           0           1  High skill tech staff   
131537                0           0      

2. Buatlah column baru bernama `has_car_and_property` dimana nilainya merupakan range dari `0-1`. `1` apabila `flag_own_car` bernilai `1` dan `flag_own_property` bernilai `1`. `0` apabila terdapat atau semua nilai pada column `flag_own_car` dan `flag_own_property` terdapat angka 0. (5pts)

In [69]:
df3_with_email['has_car_and_property'] = df3_with_email[['flag_own_car', 'flag_own_property']].apply(
    lambda row: 1 if (row['flag_own_car'] == 1) and (row['flag_own_property'] == 1) else 0, axis=1
)
df3_with_email

id  gender  flag_own_car  flag_own_property  total_children  \
5       5541414  Female             0                  1               0   
8       5169560  Female             1                  1               0   
22      5884982  Female             0                  1               0   
23      6297619    Male             1                  1               0   
28      6006818  Female             0                  1               0   
...         ...     ...           ...                ...             ...   
131511  5345596    Male             1                  1               1   
131514  5680139  Female             0                  0               1   
131518  5708288  Female             0                  1               2   
131537  5352502  Female             1                  1               0   
131555  6234604  Female             1                  1               1   

        amount_income_total           income_type  \
5                  211500.0               Working   
8                   90000.0               Working   
22                 337500.0  Commercial associate   
23                 360000.0  Commercial associate   
28                 247500.0             Pensioner   
...                     ...                   ...   
131511             112500.0               Working   
131514              67500.0               Working   
131518             135000.0               Working   
131537             103500.0  Commercial associate   
131555             171000.0               Working   

                       education_type         family_status  \
5       Secondary / secondary special        Civil marriage   
8       Secondary / secondary special               Married   
22      Secondary / secondary special               Married   
23                   Higher education  Single / not married   
28                   Higher education               Married   
...                               ...                   ...   
131511               Higher education               Married   
131514  Secondary / secondary special               Married   
131518               Higher education               Married   
131537               Higher education        Civil marriage   
131555  Secondary / secondary special               Married   

               housing_type  days_birth  days_employed  flag_mobile_phone  \
5         House / apartment      -18171          -1671                  1   
8         House / apartment      -16023          -3532                  1   
22      Municipal apartment      -15491           -703                  1   
23         Rented apartment      -15796          -1625                  1   
28        House / apartment      -23032         365243                  1   
...                     ...         ...            ...                ...   
131511    House / apartment      -17318           -246                  1   
131514    House / apartment      -15140          -2859                  1   
131518         With parents      -13296          -2178                  1   
131537    House / apartment      -11772           -113                  1   
131555     Office apartment      -13908           -993                  1   

        flag_work_phone  flag_phone  flag_email        occupation_type  \
5                     0           1           1                    NaN   
8                     0           0           1            Sales staff   
22                    0           1           1               Managers   
23                    0           0           1  High skill tech staff   
28                    0           0           1                    NaN   
...                 ...         ...         ...                    ...   
131511                0           0           1               Managers   
131514                0           0           1          Cooking staff   
131518                0           0           1  High skill tech staff   
131537                0           0      

3. Ambil column yang diperlukan yaitu `id`, `gender`, `has_car_and_property`, `income_type`, `education_type`, `family_status`, `housing_type`, `occupation_type`, `amount_income_total`. (5pts)

In [70]:
df3_dengan_kolom_yang_diperlukan = df3_with_email[['id','gender','has_car_and_property','income_type','education_type','family_status','housing_type','occupation_type','amount_income_total']]
df3_dengan_kolom_yang_diperlukan

id  gender  has_car_and_property           income_type  \
5       5541414  Female                     0               Working   
8       5169560  Female                     1               Working   
22      5884982  Female                     0  Commercial associate   
23      6297619    Male                     1  Commercial associate   
28      6006818  Female                     0             Pensioner   
...         ...     ...                   ...                   ...   
131511  5345596    Male                     1               Working   
131514  5680139  Female                     0               Working   
131518  5708288  Female                     0               Working   
131537  5352502  Female                     1  Commercial associate   
131555  6234604  Female                     1               Working   

                       education_type         family_status  \
5       Secondary / secondary special        Civil marriage   
8       Secondary / secondary special               Married   
22      Secondary / secondary special               Married   
23                   Higher education  Single / not married   
28                   Higher education               Married   
...                               ...                   ...   
131511               Higher education               Married   
131514  Secondary / secondary special               Married   
131518               Higher education               Married   
131537               Higher education        Civil marriage   
131555  Secondary / secondary special               Married   

               housing_type        occupation_type  amount_income_total  
5         House / apartment                    NaN             211500.0  
8         House / apartment            Sales staff              90000.0  
22      Municipal apartment               Managers             337500.0  
23         Rented apartment  High skill tech staff             360000.0  
28        House / apartment                    NaN             247500.0  
...                     ...                    ...                  ...  
131511    House / apartment               Managers             112500.0  
131514    House / apartment          Cooking staff              67500.0  
131518         With parents  High skill tech staff             135000.0  
131537    House / apartment            Accountants             103500.0  
131555     Office apartment            Sales staff             171000.0  

[12235 rows x 9 columns]

4. Lakukan filter dengan ketentuan sebagai berikut: (5pts)
    - Memiliki `car` dan `property`
    - Tidak ada nilai null pada column `occupation_type`
    - Memiliki value pada column `occupation_type` dengan pola `staff`.

In [71]:
df3_dengan_filter_yang_diperlukan = (
    df3_dengan_kolom_yang_diperlukan)[
    (df3_dengan_kolom_yang_diperlukan['has_car_and_property'] ==1) &
    (df3_dengan_kolom_yang_diperlukan['occupation_type'].notnull()) &
    (df3_dengan_kolom_yang_diperlukan['occupation_type'].str.lower().str.contains('staff'))
    
]
df3_dengan_filter_yang_diperlukan

id  gender  has_car_and_property           income_type  \
8       5169560  Female                     1               Working   
23      6297619    Male                     1  Commercial associate   
169     5548993  Female                     1  Commercial associate   
276     5344333  Female                     1               Working   
353     5499766  Female                     1               Working   
...         ...     ...                   ...                   ...   
131412  6046953    Male                     1         State servant   
131439  6484818    Male                     1               Working   
131470  5883006  Female                     1         State servant   
131481  6728550  Female                     1               Working   
131555  6234604  Female                     1               Working   

                       education_type         family_status  \
8       Secondary / secondary special               Married   
23                   Higher education  Single / not married   
169                  Higher education               Married   
276     Secondary / secondary special               Married   
353     Secondary / secondary special        Civil marriage   
...                               ...                   ...   
131412               Higher education  Single / not married   
131439               Higher education               Married   
131470               Higher education  Single / not married   
131481              Incomplete higher        Civil marriage   
131555  Secondary / secondary special               Married   

             housing_type        occupation_type  amount_income_total  
8       House / apartment            Sales staff              90000.0  
23       Rented apartment  High skill tech staff             360000.0  
169     House / apartment             Core staff             405000.0  
276     House / apartment         Medicine staff             135000.0  
353     House / apartment             Core staff             225000.0  
...                   ...                    ...                  ...  
131412  House / apartment             Core staff             450000.0  
131439  House / apartment            Sales staff             315000.0  
131470  House / apartment         Medicine staff             157500.0  
131481  House / apartment             Core staff             157500.0  
131555   Office apartment            Sales staff             171000.0  

[1258 rows x 9 columns]

5. Sudah dilakukan sorting berdasarkan `income_type` dari `A-Z` dan `amount_income_total` dari `besar ke kecil`. (5pts)

In [72]:
df3_sorting= df3_dengan_filter_yang_diperlukan.sort_values(['income_type','amount_income_total'], ascending=[True,False])
df3_sorting

id  gender  has_car_and_property           income_type  \
52421   6604643    Male                     1  Commercial associate   
59763   6604646    Male                     1  Commercial associate   
64362   6604645    Male                     1  Commercial associate   
24436   6423956  Female                     1  Commercial associate   
53672   6423944  Female                     1  Commercial associate   
...         ...     ...                   ...                   ...   
52545   6659184  Female                     1               Working   
52559   5021814  Female                     1               Working   
80932   5193900  Female                     1               Working   
117940  6659179  Female                     1               Working   
73293   5936326    Male                     1               Working   

                       education_type         family_status  \
52421   Secondary / secondary special  Single / not married   
59763   Secondary / secondary special  Single / not married   
64362   Secondary / secondary special  Single / not married   
24436                Higher education               Married   
53672                Higher education               Married   
...                               ...                   ...   
52545   Secondary / secondary special               Married   
52559   Secondary / secondary special               Married   
80932                Higher education               Married   
117940  Secondary / secondary special               Married   
73293   Secondary / secondary special               Married   

             housing_type occupation_type  amount_income_total  
52421   House / apartment      Core staff            1350000.0  
59763   House / apartment      Core staff            1350000.0  
64362   House / apartment      Core staff            1350000.0  
24436   House / apartment      Core staff             837000.0  
53672   House / apartment      Core staff             837000.0  
...                   ...             ...                  ...  
52545   House / apartment      Core staff              67500.0  
52559   House / apartment   Cooking staff              67500.0  
80932   House / apartment     Sales staff              67500.0  
117940  House / apartment      Core staff              67500.0  
73293   House / apartment  Security staff              40500.0  

[1258 rows x 9 columns]

6. Simpan kedalam bentuk `.csv` dengan nama `homework_3_result.csv`. (5pts)


In [73]:
df3_result = df3_sorting.to_csv('homework_3_result.csv', index=False)

# B.DataFrame Aggregation (25pts)

## Intermediate

### 4. Group By and Aggregate (10pts)

Dengan menggunakan data `titanic.csv`, buatlah dataframe untuk melihat informasi `max` dan `min` pada column `Age` dan juga informasi `mean` dan `median` pada column `Fare` pada masing-masing PassengerClass atau `Pclass`.

**Expected Result:**

![Expected Result](assets/homework_exercise_4_result.png)

In [74]:
## Your code below here
df4 = pd.read_csv('titanic.csv')
df4

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [76]:
df4_aggretiation = df4.groupby('Pclass').agg({'Age':['max','min'], 'Fare':['mean','median']})
df4_aggretiation

Age             Fare         
         max   min       mean   median
Pclass                                
1       80.0  0.92  84.154687  60.2875
2       70.0  0.67  20.662183  14.2500
3       74.0  0.42  13.675550   8.0500

### 5. Memanfaatkan Pivot Tabel (15pts)

Buatlah dataframe baru menggunakan fungsi pivot table untuk melihat `Embarked` apa saja yang `average_female_ticket_price` lebih besar daripada `average_male_ticket_price`. Informasi `average_female_ticket_price` dan `average_male_ticket_price` didapatkan dari rata-rata pada column `Fare`.

**Hint:**
1. menggunakan pivot table
2. dapat dilihat pada topik `DataFrame Aggregation` bagian `Pivot Table | Reshape Columns and Rows`

**Expected Result:**

![Expected Result](assets/homework_exercise_5_result.png)

In [126]:
## Your code below here
df5 = pd.read_csv('titanic.csv')
df5

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [128]:
embark_modif = df5[df5['Embarked'] != 'Q']
df5_grouped = df5.groupby(['Embarked','Sex']).agg({'Fare':'mean'}).reset_index()

df5_grouped



Embarked     Sex       Fare
0        C  female  75.169805
1        C    male  48.262109
2        Q  female  12.634958
3        Q    male  13.838922
4        S  female  38.740929
5        S    male  21.711996

In [130]:
df5_pivot = pd.pivot_table(
    df5_grouped,
    index=['Embarked'],
    columns=['Sex'],
    values= ['Fare']
)
df5_pivot = df5_pivot.reset_index()
df5_pivot.columns = ["Embarked","avg_female_ticket_price","avg_male_ticket_price"]
df5_pivot[df5_pivot['avg_female_ticket_price'] > df5_pivot['avg_male_ticket_price']]

Embarked  avg_female_ticket_price  avg_male_ticket_price
0        C                75.169805              48.262109
2        S                38.740929              21.711996


# C. DataFrame Combination (30pts)

## 6. Melakukan Append pada DataFrame (5pts)

Buatlah DataFrame seperti pada soal `no.1`.
 Kemudian tambahkan data tersebut dengan data users yang baru seperti gambar dibawah berikut.

**New Users Data:**

![Expected Result](assets/homework_exercise_6_new.png)

**Expected Result:**

![Expected Result](assets/homework_exercise_6_result.png)

In [137]:
## Your code below here
data = {
    "name": ["ali", "adit"],
    "age": [37,12],
    "phone_number": [None, '+62152155'],
    "ielts_score": [5.5,6.0]
}
df6 = pd.DataFrame(data)
df = df._append(df6,ignore_index=True)
df

name  age  phone_number  ielts_score
0   Fiqry   23  +62813123414          6.5
1   Iqbal   21   +6287842464          NaN
2  Monica   22  +62813125554          7.5
3    Rama   24   +6287834464          6.5
4   Johan   26  +62813113414          8.0
5     ali   37          None          5.5
6    adit   12     +62152155          6.0

### 7. Merge DataFrame (25pts)

Buatlah 2 dataframe mengikuti petunjuk berikut:

df_1 :  Filter dataset 01 telecom_revenue.csv dengan multiple kondisi (5pts)
- MonthlyRevenue > 10 
- Occupation terdiri dari Professional, Student, and Crafts

df_2 :  Filter dataset 02 telecom_usage.csv dengan multiple kondisi (5pt)
- nilai UnansweredCalls > BlockedCalls

Jawablah pertanyaan berikut ini:
1. Hitunglah total CustomerID dan rata-rata DroppedCalls untuk masing-masing occupation. (10 Points)
2. Pada occupation apa, rata-rata DroppedCalls paling besar? (5 Points)

**Expected Result:**

![Expected Result](assets/homework_exercise_7_result.png)

In [165]:
## Your code below here
df7a = pd.read_csv('01 telecom_revenue.csv')
df7a

CustomerID Churn  MonthlyRevenue  MonthlyMinutes  TotalRecurringCharge  \
0        3164846    No           34.91           619.0                  18.0   
1        3203230   Yes           43.24           405.0                  40.0   
2        3371590    No           73.47          1281.0                  71.0   
3        3138302    No          776.27          3978.0                 160.0   
4        3375774    No          189.20          1474.0                  95.0   
...          ...   ...             ...             ...                   ...   
9995     3267618    No           30.14           152.0                  40.0   
9996     3316790    No           50.61           260.0                  58.0   
9997     3265178   Yes           87.18           960.0                  85.0   
9998     3282934   Yes           31.24           134.0                  40.0   
9999     3017238    No           56.80           426.0                  30.0   

      OverageMinutes  RoamingCalls Occupation  
0               53.0           0.0      Other  
1                0.0           0.0      Other  
2                0.0           0.0      Other  
3             2278.0          22.9      Other  
4                0.0           0.0      Other  
...              ...           ...        ...  
9995             0.0           0.0      Other  
9996             0.0           0.3      Other  
9997             0.0           0.0      Other  
9998             0.0           0.0   Clerical  
9999            73.0           0.0      Other  

[10000 rows x 8 columns]

In [169]:
df7a_final= df7a[df7a['MonthlyRevenue'] > 10][df7a['Occupation'].isin(['Professional','Student','Crafts'])]
df7a_final

C:\Users\moham\AppData\Local\Temp\ipykernel_10968\1199498979.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df7a_final= df7a[df7a['MonthlyRevenue'] > 10][df7a['Occupation'].isin(['Professional','Student','Crafts'])]


CustomerID Churn  MonthlyRevenue  MonthlyMinutes  TotalRecurringCharge  \
12       3397170    No           83.51           290.0                  93.0   
25       3274822    No           30.83           273.0                  40.0   
26       3394258    No           52.72          1501.0                  62.0   
29       3056178   Yes           54.09           412.0                  40.0   
35       3065718    No           43.06           110.0                  35.0   
...          ...   ...             ...             ...                   ...   
9964     3015822    No           46.10           203.0                  38.0   
9968     3323878    No           44.26           419.0                  45.0   
9970     3050290   Yes           51.06           192.0                  30.0   
9981     3119258    No           39.49           352.0                  45.0   
9994     3125862    No           22.50            18.0                  17.0   

      OverageMinutes  RoamingCalls    Occupation  
12               0.0           0.0  Professional  
25               0.0           0.8  Professional  
26               0.0           0.0        Crafts  
29              46.0           0.0  Professional  
35               0.0           3.3        Crafts  
...              ...           ...           ...  
9964            30.0           0.0        Crafts  
9968             2.0           0.0  Professional  
9970            74.0           0.0  Professional  
9981             0.0           1.7        Crafts  
9994             2.0           0.0  Professional  

[2023 rows x 8 columns]

In [170]:
df7b = pd.read_csv('02 telecom_usage.csv')
df7b

CustomerID  DroppedCalls  BlockedCalls  UnansweredCalls  \
0        3389874           0.0           1.0              7.7   
1        3282042          15.3           1.7            152.0   
2        3054454           8.7           0.3             30.3   
3        3091842           6.0           1.7             19.3   
4        3382382           0.7           0.7              5.0   
...          ...           ...           ...              ...   
4995     3302462           6.7           0.0              4.7   
4996     3169318           0.0          43.3             23.7   
4997     3066482           0.0           3.0              3.3   
4998     3059734           0.3           0.7              2.0   
4999     3224474           6.7           1.0             32.7   

      CustomerCareCalls  
0                   0.7  
1                   5.7  
2                   0.0  
3                   0.0  
4                   0.0  
...                 ...  
4995                0.0  
4996                7.7  
4997                0.0  
4998                0.0  
4999                0.0  

[5000 rows x 5 columns]

In [171]:
df7b_final = df7b[df7b['UnansweredCalls'] > df7b['BlockedCalls']]
df7b_final

CustomerID  DroppedCalls  BlockedCalls  UnansweredCalls  \
0        3389874           0.0           1.0              7.7   
1        3282042          15.3           1.7            152.0   
2        3054454           8.7           0.3             30.3   
3        3091842           6.0           1.7             19.3   
4        3382382           0.7           0.7              5.0   
...          ...           ...           ...              ...   
4994     3148902           0.3           0.0              4.3   
4995     3302462           6.7           0.0              4.7   
4997     3066482           0.0           3.0              3.3   
4998     3059734           0.3           0.7              2.0   
4999     3224474           6.7           1.0             32.7   

      CustomerCareCalls  
0                   0.7  
1                   5.7  
2                   0.0  
3                   0.0  
4                   0.0  
...                 ...  
4994                0.0  
4995                0.0  
4997                0.0  
4998                0.0  
4999                0.0  

[4241 rows x 5 columns]

In [173]:
df7c = df7a_final.merge(df7b_final, on = 'CustomerID')
df7c

CustomerID Churn  MonthlyRevenue  MonthlyMinutes  TotalRecurringCharge  \
0       3213958   Yes           34.74           228.0                  41.0   
1       3155282    No          106.67           773.0                  80.0   
2       3326638    No           54.18           734.0                  70.0   
3       3138494    No          133.04          1716.0                  76.0   
4       3344578    No           45.63           731.0                  40.0   
..          ...   ...             ...             ...                   ...   
152     3077290    No           26.58             8.0                  25.0   
153     3389562    No           63.88           934.0                  60.0   
154     3241558   Yes           57.51           769.0                  60.0   
155     3230294    No           12.56           146.0                  10.0   
156     3029958   Yes           31.08            84.0                  30.0   

     OverageMinutes  RoamingCalls    Occupation  DroppedCalls  BlockedCalls  \
0               0.0           0.0  Professional           2.0           1.3   
1              31.0          36.4        Crafts           3.7           3.0   
2               0.0           0.0        Crafts           8.7           0.3   
3             160.0           0.0  Professional           2.0           0.0   
4              14.0           0.0  Professional          13.0           2.3   
..              ...           ...           ...           ...           ...   
152             0.0           0.0  Professional           0.3           0.3   
153            31.0           1.2       Student          26.7           0.0   
154             7.0           0.0  Professional           7.7           0.0   
155             9.0           0.0  Professional           0.3           1.0   
156             0.0           0.0  Professional           0.0           1.0   

     UnansweredCalls  CustomerCareCalls  
0               21.0                4.3  
1               24.7                0.7  
2               29.0                2.0  
3                8.3                1.0  
4               35.7               10.7  
..               ...                ...  
152              0.7                0.0  
153             26.3                0.3  
154             21.3                0.0  
155             16.3                0.0  
156             15.0                0.0  

[157 rows x 12 columns]

In [175]:
# Nomor 7.1
df7c_grouped = df7c.groupby('Occupation').agg({'CustomerID':['nunique'], 'DroppedCalls':['mean']})
df7c_grouped

CustomerID DroppedCalls
                nunique         mean
Occupation                          
Crafts               24     6.191667
Professional        128     5.532813
Student               5     8.220000

### Nomor 7.2
Pada Occupation Studentlah rata-rata DroppedCalls paling besar